### PR TITLE
Use ACE for JSON editor

### PR DESCRIPTION
**Pull request by fiftin**
Please accept PR ```ace``` to ```semamphore-vendor``` before!
Submodules update required.

### CODE REVIEW

**Comment by twhiston** [comment]
I'll try to test this over the weekend and get back to you with any feedback, thanks very much for the effort.

**Comment by twhiston** [comment]
I've just been checking this, a few points:

- [ ] Syntax checking is not activated for Environments
- [ ]  Ace does not show the gutter or syntax checking for the New Task Template 'Extra CLI Arguments' field
- [ ]  Ace does not show the gutter or syntax checking for Task run environment overrides
- [ ] Syntax checking for Inventory is incorrect, looks like it is trying to validate json but it should be validating ini or nothing

**Comment by fiftin** [comment]
Fixed. Added **ini** syntax to ```ansible-vendor``` PR.

**Comment by fiftin** [comment]
I'm not found ini-syntax validator for ACE

**Comment by fiftin** [comment]
Please accept PR to ```semaphore-vendor``` it only adds ACE library

File: public/html/projects/templates/add.pug [hunk 1]
```diff
@@ -44,9 +44,10 @@
 
 		hr
 		.form-group
-			label.control-label.col-sm-4 Extra CLI Arguments
+			label.control-label.col-sm-4 Extra CLI Arguments (*MUST* be a JSON array!)
 			.col-sm-6
-				textarea.form-control(placeholder='*MUST* be a JSON array! Each argument must be an element of the array, for example: ["-i", "@myinventory.sh", "--private-key=/there/id_rsa", "-vvvv"]' ng-model="tpl.arguments" rows="5")
+				div(ui-ace="{mode: 'json', workerPath: '/public/js/ace/'}" style="height: 100px" class="form-control" ng-model="task.environment")
+				//textarea.form-control(placeholder='*MUST* be a JSON array! Each argument must be an element of the array, for example: ["-i", "@myinventory.sh", "--private-key=/there/id_rsa", "-vvvv"]' ng-model="tpl.arguments" rows="5")
```
Comment (twhiston) [inline public/html/projects/templates/add.pug]: is there any reason to keep this commented out line?

File: public/html/projects/environment/add.pug [hunk 2]
```diff
@@ -9,7 +9,8 @@
 		input.form-control(type="text" ng-model="env.name" placeholder="Friendly name to identify your environment")
 
 		label.control-label Environment
-		textarea.form-control(ng-model="env.json" rows="20" placeholder="This has to be a JSON object")
+		div(ui-ace="{mode: 'json', workerPath: '/public/js/ace/'}" class="form-control" style="height: 200px" ng-model="env.json")
+		//textarea.form-control(ng-model="env.json" rows="20" placeholder="This has to be a JSON object")
```
Comment (twhiston) [inline public/html/projects/environment/add.pug]: is there any reason to keep this commented out line?

File: public/bundle.json [hunk 3]
```diff
@@ -6,6 +6,11 @@
 	"/vendor/angular-loading-bar/loading-bar",
 	"/vendor/moment/moment.min",
 	"/vendor/sweetalert/sweetalert.min",
+    "/vendor/ace/ui-ace.min",
+    "/vendor/ace/ace.min",
+	"/vendor/ace/mode-json",
```
Comment (twhiston) [inline public/bundle.json]: please keep the indentation consistent here

**Comment by twhiston** [comment]
PR accepted in the vendor repo

File: public/html/projects/environment/add.pug (hunk 2 above)
Comment (fiftin) [reply public/html/projects/environment/add.pug]: ACE editor has no placeholder. I kept this line to preserve placeholder. Perhaps anybody want move this message to label or to tooltip.

**Comment by fiftin** [comment]
Thank you

File: public/html/projects/environment/add.pug (hunk 2 above)
Comment (twhiston) [reply public/html/projects/environment/add.pug]: How about a workaround like this https://stackoverflow.com/questions/26695708/how-can-i-add-placeholder-text-when-the-editor-is-empty

If not can you just add another comment line above explaining that we keep this incase we want to restore the placeholder later. Thanks!

**Comment by twhiston** [comment]
Is this good to go now in your opinion? It's looking good to me :grin:

**Comment by fiftin** [comment]
I think yes :)

**Comment by twhiston** [comment]
Merged, thanks for all your work on this feature

**Pull request merged**